lib/stringlib/matching.go

```diff
@@ -318,48 +325,60 @@ func gsub(t *rt.Thread, c *rt.GoCont) (rt.Cont, *rt.Error) {
 		gc := captures[0]
 		start, end := gc.Start(), gc.End()
 		if allowEmpty || start != si || end != si {
-			sub, err := replF(captures)
+			sub, same, err := replF(captures)
 			if err != nil {
 				return nil, err
 			}
-			t.RequireBytes(si - start)
-			// No need to require memory for sub as that has been done already
-			// by replF
-			_, _ = sb.WriteString(s[si:start])
-			_, _ = sb.WriteString(sub)
+			if !same {
+				t.RequireBytes(start - sj)
+				// No need to require memory for sub as that has been done already
+				// by replF
+				_, _ = sb.WriteString(s[sj:start])
+				_, _ = sb.WriteString(sub)
+				sj = end
+			}
 		}
 		allowEmpty = start >= end
 		if allowEmpty {
-			if start < len(s) {
-				t.RequireBytes(1)
-				_ = sb.WriteByte(s[start])
-			}
+			// if start < len(s) {
+			// 	t.RequireBytes(1)
+			// 	_ = sb.WriteByte(s[start])
+			// 	sj += 1
+			// }
```

Comment:
did you wanna remove this part? :point_up:

Reply:
Yes! done.